### PR TITLE
feat(user-pool): add a UserPool Construct with type-safe triggers

### PR DESCRIPTION
**Pull request by sam-goodwin**
Implements AWS Cognito User Pools for #418

Adds a new UserPool Construct that is a light wrapper around `aws_cogntito.UserPool`, adding typed  Triggers using Functionless's `Function`.

There are three ways to configure a Trigger:

1. pass a [Function](./function/index.md) to the `lambdaTriggers` property when instantiating the `UserPool`.

```ts
// option 1 - in-line the Function
const userPool = new UserPool(stack, "UserPool", {
  lambdaTriggers: {
    createAuthChallenge: new Function(
      stack,
      "CreateAuthChallenge",
      async (event) => {
        // implement logic for the CreateAuthChallenge lifecycle event
        return event;
      }
    ),
  },
});
```

2. Call `userPool.on`.

```ts
userPool.on(
  "createAuthChallenge",
  new Function(stack, "CreateAuthChallenge", async (event) => {
    // implement logic for the CreateAuthChallenge lifecycle event
    return event;
  })
);
```

3. Call the specific `userPool.onX` method:

```ts
userPool.onCreateAuthChallenge(
  new Function(stack, "CreateAuthChallenge", async (event) => {
    // implement logic for the CreateAuthChallenge lifecycle event
    return event;
  })
);
```

### CODE REVIEW

File: src/user-pool.ts [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+  readonly customEmailSender?: Trigger<lambda.CustomEmailSenderTriggerEvent>;
+  /**
+   * Amazon Cognito invokes this trigger to send SMS notifications to users.
+   * @see https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-custom-sms-sender.html
+   * @default - no trigger configured
+   */
+  readonly customSmsSender?: Trigger<lambda.CustomSMSSenderTriggerEvent>;
+}
+
+// map a friendly trigger name to the {@link aws_cognito.UserPoolOperation}.
```
Comment (thantos) [inline src/user-pool.ts:88]: nit use block comment to get `@link`

File: src/user-pool.ts [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+                ]
+              )
+            )
+          : undefined,
+      });
+    }
+  }
+
+  /**
+   * Configure a {@link Function} to handle AWS Cognito Triggers.
```
Comment (thantos) [inline src/user-pool.ts:174]: Nit: `Configure AWS Cognito to trigger a {@link Function}.` (The function is usable in theory).

File: src/user-pool.ts [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+  readonly resource: aws_cognito.UserPool;
+
+  /**
+   * Create a new {@link UserPool} from an existing {@link aws_cognito.UserPool}.
+   *
+   *
+   * @param resource an existing UserPool Construct.
+   * @see {@link aws_cognito.UserPool}
+   */
+  constructor(resource: aws_cognito.UserPool);
```
Comment (thantos) [inline src/user-pool.ts:136]: All of the import cases should use a "from" method. #134

File: src/util.ts [hunk 4]
```diff
@@ -27,6 +27,7 @@ import {
 } from "./guards";
 import { FunctionlessNode } from "./node";
 
+export type Mixin<T> = T;
```
Comment (thantos) [inline src/util.ts:30]: What does this do?

**Review by thantos — CHANGES_REQUESTED** [review]

File: src/user-pool.ts (hunk 3 above)
Comment (sam-goodwin) [reply src/user-pool.ts:136]: I can add the `from` method but am I forced to remove the constructor? I don't need a `BaseUserPool`.

File: src/user-pool.ts [hunk 5]
```diff
@@ -190,44 +333,240 @@ export class UserPool {
...
+   * ```
+   *
+   * @see https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-custom-message.html
+   * @default - no trigger configured
+   */
+  public onCustomMessage(
+    handler: Trigger<lambda.CustomMessageTriggerEvent>
+  ): void {
+    return this.on("customMessage", handler);
+  }
```
Comment (thantos) [inline src/user-pool.ts:393]: Why did you need to get rid of the dynamic ones? Just curious, fine with explicit.

File: src/user-pool.ts (hunk 3 above)
Comment (thantos) [reply src/user-pool.ts:277]: I guess we have been removing them, shouldn't we be consistent? The harm would be that multiple constructor overrides would get messy or if there are non-import only methods, it would be harder to model that.

For example StepFunctions exposes `definition`, unless its an imported Sfn. 

Maybe its more of an issue for transpiled resources.

The argument for keeping the constructor is to avoid the base class and inheritance?

File: src/user-pool.ts [hunk 6]
```diff
@@ -0,0 +1,572 @@
...
+  defineAuthChallenge: aws_cognito.UserPoolOperation.DEFINE_AUTH_CHALLENGE,
+  postAuthentication: aws_cognito.UserPoolOperation.POST_AUTHENTICATION,
+  postConfirmation: aws_cognito.UserPoolOperation.POST_CONFIRMATION,
+  preAuthentication: aws_cognito.UserPoolOperation.PRE_AUTHENTICATION,
+  preSignUp: aws_cognito.UserPoolOperation.PRE_SIGN_UP,
+  preTokenGeneration: aws_cognito.UserPoolOperation.PRE_TOKEN_GENERATION,
+  userMigration: aws_cognito.UserPoolOperation.USER_MIGRATION,
+  verifyAuthChallengeResponse:
+    aws_cognito.UserPoolOperation.VERIFY_AUTH_CHALLENGE_RESPONSE,
+};
```
Comment (thantos) [inline src/user-pool.ts:236]: Is there an escape hatch when a new trigger name is added? (idk if this happens).

Could we accept the actual value behind the CDK enum instead of remapping a string to the enum? That way a user could override and pass in any name if a new one is added.

File: src/user-pool.ts (hunk 5 above)
Comment (sam-goodwin) [reply src/user-pool.ts:393]: I was worried it would confuse developers and we can't expand on doc and typedoc is probably terrible

Comment (sam-goodwin) [reply src/user-pool.ts:393]: Although I was particularly proud of that. It felt very much like a macro lol

File: src/user-pool.ts [hunk 7]
```diff
@@ -0,0 +1,572 @@
...
+   * @returns a Functionless {@link UserPool}.
+   */
+  public static from(resource: aws_cognito.UserPool): UserPool {
+    return new UserPool(resource);
+  }
+
+  /**
+   * The underlying {@link aws_cognito.UserPool} instance.
+   */
+  readonly resource: aws_cognito.UserPool;
```
Comment (thantos) [inline src/user-pool.ts:268]: Was going to ask if this should be `IUserPool`, but I see that `addTrigger` is on `UserPool` only. Thats annoying. I bet imported user pools cannot have triggers. 

Do we have a case where we'd want a user pool to be wrapped even if imported? (I think so?)

Comment (thantos) [reply src/user-pool.ts:268]: https://github.com/aws/aws-cdk/blob/v1.169.0/packages/@aws-cdk/aws-cognito/lib/user-pool.ts#L708

File: src/user-pool.ts (hunk 3 above)
Comment (sam-goodwin) [reply src/user-pool.ts:277]: Mostly to avoid an extra class but I am fine with changing it to be consistent. I usually hate Base* classes but I also get the point off it confusing developers. The first constructor they see is the import one so that could be a problem. I can roll in a fix for secrets into this change.

File: src/user-pool.ts (hunk 7 above)
Comment (thantos) [reply src/user-pool.ts:268]:
```ts
functionless.UserPool.from(aws_cognito.UserPool.fromUserPoolArn(...))
```

or

```ts
functionless.UserPool.fromUserPoolArn(...)
```

Comment (thantos) [reply src/user-pool.ts:268]: That said... step function relies on `StepFunction` not `IStepFunction` because `IStepFunction` doesn't support `stateMachineType` and thus cannot be validated... maybe we need to fix that.

**Review by thantos — APPROVED** [review]

File: src/user-pool.ts (hunk 6 above)
Comment (sam-goodwin) [reply src/user-pool.ts:236]: Ok, made it possible al ways around.

```ts
test("add a trigger to a user pool using on(aws_cognito.UserPoolOperation)", () => {
  const stack = new Stack();

  const pool = new UserPool(stack, "UserPool");

  pool.on(
    aws_cognito.UserPoolOperation.DEFINE_AUTH_CHALLENGE,
    new Function(
      stack,
      "CreateAuthChallenge",
      // must provide explicit event type
      async (event: CreateAuthChallengeTriggerEvent) => {
        return event;
      }
    )
  );
});

```

Comment (sam-goodwin) [reply src/user-pool.ts:236]: This has the drawback of not inferring the type of the event. The enum is a synthetic enum where the enum items are objects (they are all the same type and can't be used as keys). So the user must annotate the `event: CreateAuthChallengeTriggerEvent ` type.

File: src/user-pool.ts (hunk 5 above)
Comment (thantos) [reply src/user-pool.ts:368]: Yeah, it was cool, but agree about explicit being a better experience.

**Comment by thantos** [comment]
Looks good, just missing the change to Secret as a mentioned: https://github.com/functionless/functionless/pull/419#discussion_r949785586
